dev-docs/source/DesignDocumentGuides.rst

```diff
@@ -10,39 +10,96 @@ Design Document Guidelines
+One should always do design work before coding, no matter the scope and the nature of the feature.
+These guidelines are to answer in which cases the design must be a *public*, *written* and *persistent* document, rather than a sketch on a white-board in a stand-up meeting that is wiped-out after it has been turned into code.
+Just as coding, design is an iterative and collaborative process.
+Having the document written allows for discussion and iterations with peers before the first line of code is written.
+Besides, the written document facilitates the pull request review process.
+Having the design already approved, the reviewer will not need to do a design review with the code review, but will just need to make sure that the implementation matches the previously agreed design.
+Finally, once written, the document can also serve as (or easily be turned into) a developer documentation for future reference.
+
+Not every feature requires a design document, as the benefit it adds must be gaged with the time and effort that goes into it.
+Bug fixes, small extensions to the current functionalities, as well as concrete, single-class plugins, such as algorithms, do not need a document.
+In the contrary, if the feature is large-scope, brand-new, many-class, including abstractions, and/or touches or refactors the internals of the core framework, it will most likely need a document.
+Below are a couple of metrics that could help a developer decide whether or not the feature falls into that category.
+
+- Number of classes and components.
+
+  If the number of classes is more than just a few, the document is relevant in order to describe the relations between them.
+
+- Fraction of abstract classes and interfaces in the design.
+
+  Abstractions and APIs are hard to change, once in. So whenever the feature concerns creations of new abstraction layers and APIs, a wider discussion on the design is needed, hence the necessity of the document.
+
+- Fraction of files and components the feature will be touching.
+
+  Single file edits are less likely to require a document.
+
+- The layer the feature will be added to.
+
+  If the feature is deep in the framework or workbench, it is more critical to have a document.
+  If the feature is just a plugin (algorithm, function, GUI, script) in the periphery, with low risk of side effects, design document is less important.
+
+- Number of developers involved.
+
+  If the feature is large enough that the implementation will be done in a distributed way, the document will ensure that the whole team has a common vision, discussed and agreed upfront.
+
+- Number of users the feature will be facing.
+
+  If the feature is generic, it will be facing a wider audience of users, hence a document and design review will make sure that the right feature is being built in the right way.
+  In the contrary, if the feature is specific to one instrument or technique at one facility, producing a document is less critical.
```

Comment:
Can we say it may be less critical, yet it needs to be balanced with number of users? I'm assuming even if it was only one instrument or one technique if it has a high volume use we may need to consider a design document as it does affect a larger number of users.

Reply:
The point is that if it is a single instrument, no matter the number of end users using it, it will be a very small number compared to the overall workbench users. So it is not really the number of users of the instrument that counts (as compared to other instruments). What makes the difference is how generic the feature is. If it is a workflow for one instrument at one facility, it can't be compared with the generic tool as the workbench, even if that one isntrument has many users. So in my opinion, less critical or less important is really the best word here.

Reply:
Thats fair enough - I withdraw my comment :)